src/Search.f90

```diff
@@ -11,95 +11,87 @@
 INTERFACE lowerBound
-  MODULE PROCEDURE lower_bound_i
-  MODULE PROCEDURE lower_bound_r
+  !> @copybrief Search::lowerBound_i
+  !> @copydetails Search::lowerBound_i
+  MODULE PROCEDURE lowerBound_i
+  !> @copybrief Search::lowerBound_r
+  !> @copydetails Search::lowerBound_r
+  MODULE PROCEDURE lowerBound_r
 ENDINTERFACE lowerBound
 
-!> @brief Return the index of an element in a list which compares equal to the
-!> passed value.
-!> @param list the list to search. Unlike for binarySearch, the list need not
-!> be sorted.
-!> @param val the value to find
-!> @param tol the comparison tolerance to use for floating-point
-!> implementations. Optional
+!> @brief Interface for finding entry of a 1D array using a linear search
 !>
-!> If the value is not found, return SIZE+1
 INTERFACE find
-  MODULE PROCEDURE find_i
-  MODULE PROCEDURE find_r
+  !> @copybrief Search::linearSearch_i
+  !> @copydetails Search::linearSearch_i
+  MODULE PROCEDURE linearSearch_i
+  !> @copybrief Search::linearSearch_r
+  !> @copydetails Search::linearSearch_r
+  MODULE PROCEDURE linearSearch_r
 ENDINTERFACE find
 
-!> @brief Return the index of an element in a list which compares equal to the
-!> passed value.
-!> @param list the list to search. The list should be sorted, otherwise
-!> undefined behavior
-!> @param val the value to find
-!> @param tol the comparison tolerance to use for floating-point
-!> implementations. Optional
+!> @brief Interface for finding entry of a 1D array using a binary search
 !>
-!> If the value is not found, return SIZE+1
 INTERFACE binarySearch
-  MODULE PROCEDURE binary_search_i
-  MODULE PROCEDURE binary_search_r
+  !> @copybrief Search::binarySearch_i
+  !> @copydetails Search::binarySearch_i
+  MODULE PROCEDURE binarySearch_i
+  !> @copybrief Search::binarySearch_r
+  !> @copydetails Search::binarySearch_r
+  MODULE PROCEDURE binarySearch_r
 ENDINTERFACE binarySearch
-
 !
 !===============================================================================
 CONTAINS
 !
 !-------------------------------------------------------------------------------
-PURE FUNCTION upper_bound_i(list,val) RESULT (i)
+!> @brief Returns the index of the first element of a list of integers that is
+!>        greater than val
+!> @param list 1D array of integers to search through. The list should be sorted in
+!>        ascending order
+!> @param val integer value to find the first entry of list that is greater than
+!> @returns i the index of the first element of list that is greater than val
+!>
+!> NOTE: If val is greater than or equal to the largest element of list this
+!>       returns SIZE(list)+1, and if val is smaller than the smallest element of
+!>       list this returns 1.
+!>
+PURE FUNCTION upperBound_i(list,val) RESULT (i)
   INTEGER(SIK),INTENT(IN) :: list(:)
   INTEGER(SIK),INTENT(IN) :: val
   INTEGER(SIK) :: j
   !
-  INTEGER(SIK) :: n, step, i
+  INTEGER(SIK) :: n,step,i
 
```

Comment:
I think basically all of the functions in this module are missing some potential DBC opportunities.  You've added these `isMonotonic`, `isDecreasing`, and `isIncreasing` functions.  I think they should see some use throughout this module.

Reply:
Basically add a couple `REQUIRE(isMonotonic(array))` in the getFirstGreater/getFirstGreaterEqual and I'm good with it